bindings/python/src/lib.rs

```diff
@@ -441,9 +458,166 @@ struct Open {
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    fn read_odirect(filename: impl AsRef<Path>) -> PyResult<DirectIo> {
+        let filename = filename.as_ref().to_str().ok_or_else(|| {
+            SafetensorError::new_err(format!(
+                "Path {} is not valid UTF-8",
+                filename.as_ref().display()
+            ))
+        })?;
+
+        unsafe {
+            let c_path = std::ffi::CString::new(filename)
+                .map_err(|_| SafetensorError::new_err("Failed converting file path to CString"))?;
+            let fd = libc::open(c_path.as_ptr(), libc::O_RDONLY | libc::O_DIRECT);
+            if fd < 0 {
+                return Err(SafetensorError::new_err(format!(
+                    "Could not open file {filename} with O_DIRECT",
+                )));
+            }
+            let file = File::from_raw_fd(fd);
+            let meta = file.metadata().map_err(|_| {
+                SafetensorError::new_err(
+                    format!("Could not retrieve metadata for file {filename}",),
+                )
+            })?;
+            let file_len = meta.len() as usize;
+            std::mem::forget(file);
+            if file_len == 0 {
+                libc::close(fd);
+                return Ok(DirectIo::from_vec(vec![]));
+            }
+
+            let aligned_prefix_len = file_len / ALIGN * ALIGN;
+            let tail_len = file_len - aligned_prefix_len;
+
+            let aligned_capacity = file_len.checked_next_multiple_of(ALIGN).ok_or_else(|| {
+                SafetensorError::new_err("overflow while computing aligned file len")
+            })?;
+            let layout = Layout::from_size_align(aligned_capacity, ALIGN)
+                .map_err(|_| SafetensorError::new_err("invalid layout"))?;
+            let ptr = alloc(layout);
+            if ptr.is_null() {
+                handle_alloc_error(layout);
+            }
+            let mut vec: Vec<u8> = Vec::from_raw_parts(ptr, 0, aligned_capacity);
+
+            if aligned_prefix_len > 0 {
+                // Cap max thread count to 8, we don't need too many to saturate disk
+                let n_threads = std::thread::available_parallelism()
+                    .map(|n| n.get())
+                    .unwrap_or(1)
+                    .min(8);
+
+                let block_size = ALIGN;
+                let total_blocks = aligned_prefix_len / block_size;
+                let blocks_per_thread = total_blocks.div_ceil(n_threads);
+                let region_bytes = blocks_per_thread * block_size;
+
+                thread::scope(|scope| -> PyResult<()> {
+                    println!("creating buffer slice of len {aligned_prefix_len}");
+                    let buffer: &mut [u8] =
+                        std::slice::from_raw_parts_mut(vec.as_mut_ptr(), aligned_prefix_len);
+
+                    let mut handles = Vec::new();
+
+                    for (tid, chunk) in buffer.chunks_mut(region_bytes).enumerate() {
```

Comment:
Question for understanding, since `region_bytes = blocks_per_thread * block_size`. Suppose we have 16 blocks in the file and 4 threads. Then the responsibility by thread is:

```
0 0 0 0 1 1 1 1 2 2 2 2 3 3 3 3
```

I am curious how this compares to patterns like:

```
0 1 2 3 0 1 2 3 0 1 2 3 0 1 2 3 
```